aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs

```diff
@@ -0,0 +1,606 @@
+/// `Aggregator` Move struct.
+const TABLE_HANDLE_FIELD_INDEX: usize = 0;
+const KEY_FIELD_INDEX: usize = 1;
+const LIMIT_FIELD_INDEX: usize = 2;
+
+/// Given a reference to `AggregatorTable` Move struct, returns the value of
+/// `table_handle` field.
+fn get_table_handle(aggregator_table: &StructRef) -> PartialVMResult<TableHandle> {
+    aggregator_table
+        .borrow_field(TABLE_FIELD_INDEX)?
+        .value_as::<StructRef>()?
+        .borrow_field(HANDLE_FIELD_INDEX)?
+        .value_as::<Reference>()?
+        .read_ref()?
+        .value_as::<u128>()
+        .map(TableHandle)
+}
+
+/// Given a reference to `Aggregator` Move struct and field index, returns
+/// its field specified by `index`.
+fn get_aggregator_field(aggregator: &StructRef, index: usize) -> PartialVMResult<Value> {
+    let field_ref = aggregator.borrow_field(index)?.value_as::<Reference>()?;
+    field_ref.read_ref()
+}
+
+///  Given a reference to `Aggregator` Move struct, returns a tuple of its
+///  fields: (`table_handle`, `key`, `limit`).
+fn get_aggregator_fields(aggregator: &StructRef) -> PartialVMResult<(TableHandle, u128, u128)> {
+    let table_handle = get_aggregator_field(aggregator, TABLE_HANDLE_FIELD_INDEX)?
+        .value_as::<u128>()
+        .map(TableHandle)?;
+    let key = get_aggregator_field(aggregator, KEY_FIELD_INDEX)?.value_as::<u128>()?;
+    let limit = get_aggregator_field(aggregator, LIMIT_FIELD_INDEX)?.value_as::<u128>()?;
+    Ok((table_handle, key, limit))
+}
+
+/// Returns partial VM error on abort.
+fn abort_error(message: impl ToString, code: u64) -> PartialVMError {
+    PartialVMError::new(StatusCode::ABORTED)
+        .with_message(message.to_string())
+        .with_sub_status(code)
+}
+
+/// Returns partial VM error on extension failure.
+fn extension_error(message: impl ToString) -> PartialVMError {
+    PartialVMError::new(StatusCode::VM_EXTENSION_ERROR).with_message(message.to_string())
+}
+
+// ================================= Tests =================================
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use claim::{assert_matches, assert_none};
+    use move_deps::move_vm_test_utils::BlankStorage;
+    use once_cell::sync::Lazy;
+
+    static DUMMY_RESOLVER: Lazy<BlankStorage> = Lazy::new(|| BlankStorage);
+    static DUMMY_HANDLE: Lazy<TableHandle> = Lazy::new(|| TableHandle(0));
+
+    fn set_up(context: &NativeAggregatorContext) {
+        let mut aggregator_table_data = context.aggregator_table_data.borrow_mut();
+        let table = aggregator_table_data.get_or_create_aggregator_table(*DUMMY_HANDLE);
+
+        // Aggregators with data.
+        table.create_new_aggregator(0, 1000);
+        table.create_new_aggregator(1, 1000);
+        table.create_new_aggregator(2, 1000);
+
+        // Aggregators with delta.
+        table.get_or_create_aggregator(3, 1000);
+        table.get_or_create_aggregator(4, 1000);
+        table.get_or_create_aggregator(5, 10);
+
```

Comment:
Can we also test adding / materializing some of these? either in the set_up and same test, or another test that follows set-up with some accesses to those aggregators. Could also test getting smt flow that already exists.